### PR TITLE
[Sema] Highlight angle brackets <…> in diagnostics

### PR DESCRIPTION
**Pull request by JaSpa**
Move the checks for invalid generic arguments into `applyGenericTypeReprArgs` so we can pass the `GenericIdentTypeRepr` into it to access the angle bracket range while `resolveDictionaryType` can still use `applyUnboundGenericArguments` without an `GenericIdentTypeRepr`.

Edit: `check-swift` passed

### CODE REVIEW

File: lib/Sema/TypeChecker.h [hunk 1]
```diff
@@ -714,12 +714,12 @@ class TypeChecker final : public LazyResolver {
   ///
   /// \returns A BoundGenericType bound to the given arguments, or null on
   /// error.
-  Type applyGenericArguments(Type type,
-                             SourceLoc loc,
-                             DeclContext *dc,
-                             MutableArrayRef<TypeLoc> genericArgs,
-                             bool isGenericSignature,
-                             GenericTypeResolver *resolver);
+  Type applyUnboundGenericArguments(UnboundGenericType *unbound,
```
Comment (JaSpa) [inline lib/Sema/TypeChecker.h]: Is it ok to pass an `UnboundGenericType` or should it be wrapped inside a `Type` value?

**Comment by JaSpa** [comment]
@DougGregor would you mind taking a look?

File: lib/Sema/TypeChecker.h [hunk 2]
```diff
@@ -702,25 +702,54 @@ class TypeChecker final : public LazyResolver {
...
+  ///
+  /// \see applyUnboundGenericArguments
+  Type applyGenericArguments(Type type, SourceLoc loc, DeclContext *dc,
+                             GenericIdentTypeRepr *generic,
                              bool isGenericSignature,
                              GenericTypeResolver *resolver);
 
+  /// Apply generic arguments to the given type.
+  ///
+  /// This functions requires a valid unbound generic type with the correct
```
Comment (jpsim) [inline lib/Sema/TypeChecker.h]: function

**Comment by DougGregor** [comment]
I'd usually grumble about trailing whitespace changes, but it's fine; we should clean those up anyway. Thanks!

**Comment by DougGregor** [comment]
Oh, I suspect this means that Sema could properly distinguish between "X<>" and "X" now, perhaps with some minor parser tweaks.

**Pull request merged**